### DISCUSSION TITLE
Kitty key binding override Neovim shortcut

### DISCUSSION THREAD

**Discussion by VisionaryAppDev** [kovidgoyal/kitty#6054] [Q&A] [answered]
In most of my workflow, I often set a shortcut that switch between tabs. In most of the editor or IDE, I always bind it to alt-1, alt-2 and so on. Now, I have switched to Kitty and have tried mapping this shortcut to kitty so I can switch between tab easily and faster. But the thing is, when I bind those shortcut to kitty, Neovim is no longer execute anything when I press alt-1 or alt-2 and only Kitty is taking the action. Even when I open only a single terminal tab, it still doesn't work.

So, I just wonder if there is any solution to solve this problem? For example, dynamically send an unbind command from Neovim to Kitty?

**Comment by page-down** [comment, marked as answer]
> ... switch between tab  ... Even when I open only a single terminal tab, ...

Write a kitten to perform different actions depending on the conditions.
For example, when there is only one tab, send the key escape sequence.

https://sw.kovidgoyal.net/kitty/kittens/custom/#using-kittens-to-script-kitty-without-any-terminal-ui
There are some examples in GitHub Discussion.

For how to send the keys you can refer to the following project.
https://github.com/knubie/vim-kitty-navigator

> ... dynamically send an unbind command ...

You can also execute kitten by remote control to do whatever you want.
However, changing keys dynamically will generally cause trouble and make it difficult to troubleshoot problems, and is less recommended.

**Reply by VisionaryAppDev** [reply]
I have tried following the code you have mentioned but seem like it send the shortcut as a string to the terminal instead of shortcut. 

- kitty.conf
```
...
map ctrl+k kitten pass_keys.py
```


- pass_key.py
```
from kittens.tui.handler import result_handler
from kitty.key_encoding import KeyEvent, parse_shortcut


def encode_key_mapping(window, key_mapping):
    mods, key = parse_shortcut(key_mapping)
    event = KeyEvent(
        mods=mods,
        key=key,
        shift=bool(mods & 1),
        alt=bool(mods & 2),
        ctrl=bool(mods & 4),
        super=bool(mods & 8),
        hyper=bool(mods & 16),
        meta=bool(mods & 32),
    ).as_window_system_event()

    return window.encoded_key(event)


def main():
    pass


@result_handler(no_ui=True)
def handle_result(args, result, target_window_id, boss):
    window = boss.window_id_map.get(target_window_id)


    encoded = encode_key_mapping(window, "ctrl+shift+t")
    window.write_to_child(encoded)
    print(encoded)
```

From the debug windows it print the value as: ```b'\x1b[116;6u'``` while in the termal: ```6;6u```

So, could you guide me how to achieve the way of sending escape key or shortcut to the kitty?

**Reply by page-down** [reply]
Not sure what you mean, this works as expected.

- Map the key in Neovim: `<C-S-T>`
- kitty.conf: `map ctrl+shift+t kitten mykitten.py`
- Press the shortcut and you will see the mapped key bindings being handled in neovim.
- Modify the kitten to:
  - If there is only one tab, send the key escape sequence to the child, otherwise do something else.
- Create two tabs and press the shortcut again to see the difference.

**Reply by VisionaryAppDev** [reply]
Aww, I was confused. From the code above, I expect Kitty to create another tab when I press `ALT+1` because `ctrl+shift+t` is bind to create a new tab. But, in fact, it sends some value as I mentioned above into the Terminal instead. The thing is, when I open Neovim, it can read those value as a shortcut which is awesome. 


Now I have got everything working so I'll just leave the complete work here so in case someone find it useful in the future.

- ~/.config/kitty/kitty.conf

```
map alt+1 kitten pass_keys.py 1
map alt+2 kitten pass_keys.py 2
map alt+3 kitten pass_keys.py 3
map alt+4 kitten pass_keys.py 4
map alt+5 kitten pass_keys.py 5
map alt+6 kitten pass_keys.py 6
map alt+7 kitten pass_keys.py 7
map alt+8 kitten pass_keys.py 8
```


- ~/.config/kitty/pass_keys.py
```
from kittens.tui.handler import result_handler
from kitty.key_encoding import KeyEvent, parse_shortcut
import psutil


def encode_key_mapping(window, key_mapping):
    mods, key = parse_shortcut(key_mapping)
    event = KeyEvent(
        mods=mods,
        key=key,
        shift=bool(mods & 1),
        alt=bool(mods & 2),
        ctrl=bool(mods & 4),
        super=bool(mods & 8),
        hyper=bool(mods & 16),
        meta=bool(mods & 32),
    ).as_window_system_event()

    return window.encoded_key(event)


def main():
    pass


def is_neovim_running():
    for process in psutil.process_iter(['pid', 'name']):
        if process.info['name'] == 'nvim':
            return True
    return False


@result_handler(no_ui=True)
def handle_result(args, result, target_window_id, boss):
    window = boss.window_id_map.get(target_window_id)
    tab_to_switch = args[1] # Get args [pass_keys.py, 1] that we pass in (map alt+1 kitten pass_keys.py 1)

    if is_neovim_running():
        encoded = encode_key_mapping(window, "alt+" + tab_to_switch)
        window.write_to_child(encoded)
    else:
        # If total tab is greater than request tab, then switch, otherwise no.
        if len(boss.active_tab_manager) >= int(tab_to_switch):
            boss.goto_tab(int(tab_to_switch))
```

Btw, thanks @page-down for you help ;)

**Comment by danielb2** [comment]
I have this exact same thing with neovim

```
" alt+1 goes to tab 1, alt+2 tab 2, etc                                                                                                                                                                                                         
noremap <a-1> 1gt                                                                                                                                                                                                                               
noremap <a-2> 2gt                                                                                                                                                                                                                               
noremap <a-3> 3gt                                                                                                                                                                                                                               
noremap <a-4> 4gt                                                                                                                                                                                                                               
noremap <a-5> 5gt              
```

this pass key code is the final solution? Why doesn't the terminal actually forward keypress ?

**Reply by VisionaryAppDev** [reply]
Yes, the terminal does forward the key press if you follow the script I put above. And you need to config you neovim to switch to another buffer as well. 

Since I am using bufferline and whichkey, I have them setup as below:
```
which_key.register({
      ["<M-1>"] = { function() require('bufferline').go_to_buffer(1, true) end, "Switch to 1st tab" },
      ["<M-2>"] = { function() require('bufferline').go_to_buffer(2, true) end, "Switch to 2nd tab" },
      ["<M-3>"] = { function() require('bufferline').go_to_buffer(3, true) end, "Switch to 3rd tab" },
      ["<M-4>"] = { function() require('bufferline').go_to_buffer(4, true) end, "Switch to 4th tab" },
      ["<M-5>"] = { function() require('bufferline').go_to_buffer(5, true) end, "Switch to 5th tab" },
 )}
```

**Comment by danielb2** [comment]
I ended up just adding these to my init.vim:

```
noremap ¡ 1gt                                                                                                                                                                                                                                   
noremap ™ 2gt                                                                                                                                                                                                                                   
noremap £ 3gt                                                                                                                                                                                                                                   
noremap ¢ 4gt                                                                                                                                                                                                                                   
noremap ∞ 5gt    
```
which solved the issue for me

**Reply by kovidgoyal** [reply]
You want https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.macos_option_as_alt

**Reply by danielb2** [reply]
I had tried this but I must have been doing something wrong before. I found it later and it works now. thanks
